Wayland fixes

yol opened the pull request:
Various fixes for wayland

## Description
@wsnipex 7a04234 should fix the Ubuntu cursor issues. 2e2d90d makes resizing easier by adding some extra pixels transparent borer. You shold probably also take a look at 
152e36a.

## Motivation and Context
Some issues were discovered recently by testers

## How Has This Been Tested?
<!--- Please describe in detail how you tested your change -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc -->

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] Improvement (non-breaking change which improves existing functionality)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the [Code guidelines](https://codedocs.xyz/xbmc/xbmc/code_guidelines.html) of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the [CONTRIBUTING](https://github.com/xbmc/xbmc/blob/master/CONTRIBUTING.md) document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

Rechi commented:
@pkerling waylandpp has to be updated in depends as you require now a higher version.

pkerling commented:
Good point.

pkerling commented:
Hmmmm it's still using SCons ... I could've sworn I updated it at some point. Time to do it now then I guess.

Rechi commented on tools/depends/native/Toolchain.cmake.in:
```diff
@@ -0,0 +1,42 @@
+set(CMAKE_OBJDUMP @OBJDUMP_FOR_BUILD@ CACHE FILEPATH "Objdump")
+set(CMAKE_RANLIB @RANLIB_FOR_BUILD@ CACHE FILEPATH "Ranlib")
+
+if(PROJECT_SOURCE_DIR MATCHES "tools/depends")
```
not needed

Rechi commented on tools/depends/native/Toolchain.cmake.in:
```diff
@@ -0,0 +1,42 @@
+  endif()
+endif()
+
+# where is the target environment
```
`... the native environment`

Rechi commented on tools/depends/native/Toolchain.cmake.in:
```diff
@@ -0,0 +1,42 @@
+set(CMAKE_FIND_ROOT_PATH @prefix@/@tool_dir@)
+set(CMAKE_LIBRARY_PATH @prefix@/@tool_dir@/lib)
+if(NOT "@use_sdk_path@" STREQUAL "")
+  list(APPEND CMAKE_FIND_ROOT_PATH @use_sdk_path@ @use_sdk_path@/usr)
```
wrong, this is a target path

Rechi commented on tools/depends/native/Toolchain.cmake.in:
```diff
@@ -0,0 +1,42 @@
+if(CMAKE_SYSTEM_NAME STREQUAL Darwin)
+  set(CMAKE_OSX_SYSROOT @use_sdk_path@)
```
wrong, this is a target path

Rechi reviewed (dismissed):
depends need some adjustment

wsnipex commented:
do we really need the native toolchain file? Imho it should work fine without

Rechi commented:
> do we really need the native toolchain file?

I think we then have to pass the correct executables and flags in the Makefile

pkerling commented:
> do we really need the native toolchain file? Imho it should work fine without

It used the target toolchain file before which did not work of course, but I can try without the toolchain file at all. I guess for waylandpp scanner building it could work since it does not have any external dependencies.

Rechi commented:
reason why it used target toolchain file is https://github.com/xbmc/xbmc/blob/b1471fe7caf63df589c7f888319dd072aa0982b5/tools/depends/Makefile.include.in#L67

pkerling commented:
Let's see how it goes

pkerling commented:
I think if we build more sophisticated stuff with CMake in native it could become a problem, but for now it seems to work.

Rechi reviewed (commented):
depends changes look better now
My only concern is if native waylandpp scanner uses the correct compiler specified by `Makefile.include`.

wsnipex commented on tools/depends/native/waylandpp-scanner-native/Makefile:
```diff
@@ -0,0 +1,41 @@
+APP=$(PLATFORM)/wayland-scanner++
+CLEAN_FILES=$(ARCHIVE) $(PLATFORM)
+
+CMAKE_OPTIONS := -DCMAKE_TOOLCHAIN_FILE= -DBUILD_DOCUMENTATION=OFF -DBUILD_LIBRARIES=OFF -DBUILD_EXAMPLES=OFF -DBUILD_SCANNER=ON $(CMAKE_OPTIONS)
```
Even if makefile variable evaluation might do the right thing here, please set
CMAKE_INSTALL_PREFIX=$(NATIVEPREFIX) to be explicit.

pkerling replied:
I kind of dislike that we have to overwrite CMake arguments that are already specified (also applies to `CMAKE_TOOLCHAIN_FILE`). It will look weird in the log. Is there another way to solve this? Maybe have separate `Makefile.include` for native and target?

Rechi replied:
set `CMAKE_FOR_BUILD` (or something similar) in `Makefile.include`

wsnipex reviewed (commented):
only 1 minor

pkerling commented:
> My only concern is if native waylandpp scanner uses the correct compiler specified by Makefile.include.
You mean `CC_FOR_BUILD` etc.? I don't think it will, that's what the toolchain file would have been for.

Rechi commented:
yes, native waylandpp doesn't consider `CC_FOR_BUILD`

pkerling commented:
> yes, native waylandpp doesn't consider CC_FOR_BUILD

I see two ways to fix this (if we want to fix it):
1. Give them (and lots of other stuff) as arguments to CMake in `CMAKE_FOR_BUILD` (e.g. `-DCMAKE_C_COMPILER=...`)
1. Reintroduce the toolchain file and specify it there (I can ofc remove/fix the stuff @Rechi pointed out)

What's your opinion @Rechi @wsnipex?

wsnipex commented:
just pass CMAKE_CXX_COMPILER=$(CXX_FOR_BUILD)
As long as we have a single native lib that needs cmake, it's not worth adding a toolchain file

pkerling commented:
jenkins build this please

pkerling commented:
I can't use `CC_FOR_BUILD` because it might include `ccache` (which is not a compiler and rejected by CMake). Should I make a new variable for the plain compiler in `Makefile.include.in`?

wsnipex commented:
sure, thats fine

pkerling commented:
jenkins build this please

pkerling commented:
Jenkins is (mostly) happy now. If @wsnipex gives his OK I'll squash and merge.

Rechi commented on cmake/modules/FindWaylandpp.cmake:
```diff
@@ -22,6 +24,13 @@ if(PKG_CONFIG_FOUND)
   endif()
+
+  if(PC_WAYLANDPP_FOUND AND PC_WAYLANDPP_wayland-client++_VERSION VERSION_LESS WAYLANDPP_MINIMUM_VERSION)
+    message(FATAL_ERROR "At least version ${WAYLANDPP_MINIMUM_VERSION} of waylandpp is required, but you only have version ${PC_WAYLANDPP_wayland-client++_VERSION}")
```
should be handled by `find_package_handle_standard_args` (see FindBluray.cmake for an example)

pkerling replied:
If you use `find_package_handle_standard_args`, that means that the FindPackage call in the actual CMake file must request a `VERSION`, which is not the case.

Rechi replied:
Only `set(Waylandpp_FIND_VERSION 0.1.5)` would be needed before `find_package_handle_standard_args` call and version var has to added to that.

pkerling replied:
Not sure it's intended to be used like that, but yeah that should work :-)

Rechi replied:
We use it like that for other modules already.

pkerling replied:
done

Rechi reviewed (approved):
maybe fix the version check in cmake
cmake and depends changes look good

pkerling commented:
Ignore the last commit, I'll remove it when squashing. Won't work with non-mesa toolchains.

pkerling committed:
[wayland] Fix some off-by-ones in decoration drawing

pkerling committed:
[wayland] Draw transparent border around visible border as resize grab

pkerling committed:
Allow addition/subtraction of CSize to/from CRect

pkerling committed:
Fix indentation

pkerling committed:
[depends] Remove SCons

pkerling committed:
[wayland] Fix handling of monitors that have no useful physical size

pkerling committed:
[depends] Build waylandpp with CMake

pkerling committed:
[wayland] Support older XCURSOR themes by falling back to old names

Rechi commented:
wrong revision was build
jenkins build this please

pkerling committed:
[cmake] Bump required CMake version to 3.4

so we can rely on pkg_get_variable being available

pkerling committed:
[depends] Link xkbcommon pkgconfig file from host

xkbcommon is required for Wayland build

wsnipex commented:
I don't like the last commit. It's good practice to not trust pkg-config and always check the headers and libs. We do this almost everywhere and it proved to be necessary for some embedded stuff that plain copy libs somewhere in the toolchains.

pkerling commented:
> It's good practice to not trust pkg-config and always check the headers and libs. We do this almost everywhere and it proved to be necessary for some embedded stuff that plain copy libs somewhere in the toolchains.

What is the problem exactly with the toolchains? They have libs without providing pkgconfig files?

pkerling commented:
jenkins build this please

pkerling commented:
jenkins build this please

pkerling commented:
jenkins build this please

wsnipex commented on tools/depends/target/Toolchain.cmake.in:
```diff
@@ -105,7 +105,8 @@ set(CMAKE_FIND_ROOT_PATH_MODE_PROGRAM NEVER)
 set(CMAKE_FIND_FRAMEWORK LAST)
-set(ENV{PKG_CONFIG_LIBDIR} @prefix@/@deps_dir@/lib/pkgconfig:@prefix@/@deps_dir@/share/pkgconfig)
+# also include native pkgconfig dir to make native binaries used during build available
+set(ENV{PKG_CONFIG_LIBDIR} @prefix@/@deps_dir@/lib/pkgconfig:@prefix@/@deps_dir@/share/pkgconfig:@prefix@/@tool_dir@/lib/pkgconfig:@prefix@/@deps_dir@/share/pkgconfig)
```
nooo

pkerling replied:
We need the native binaries though?

wsnipex replied:
please define/pass them where needed. If really necessary add $(NATIVEPREFIX)/bin to path, but never globally

pkerling committed:
[cmake] Export symbols also in test executable (for cpluff)

pkerling commented:
@wsnipex I'd like to get consensus on the pkgconfig thing so I can move this forward. If we want to support no-pkgconfig setups (and broken toolchains I guess?) I have to add the peel-the-library-version-out-of-some-header-with-regex-thingy to `FindWaylandpp.cmake` and `FindXkbcommon.cmake`. Personally, I feel that not having to do this is exactly what pkgconfig is for. Regarding the `PKG_CONFIG_LIBDIR` change I wouldn't mind reverting it and making the scanner path configurable so that's a non-issue.

wsnipex commented:
re pkgconfig-only Find module: if others are fine with it, I won't block it

adding native depends to target toolchain is a no-go though

pkerling commented:
jenkins build this please
this is getting annoying

pkerling commented:
> re pkgconfig-only Find module: if others are fine with it, I won't block it

If there are valid concerns I definitely do want to hear about it. I also don't want to have to revert this first thing tomorrow because the LibreELEC guys are angry at us or something :-) @Rechi @fritsch ?

Rechi commented:
Linux isn't a platform I develop for, but as the cmake checks are already there present (`find_path` & `find_library`) I wouldn't remove them.

pkerling commented:
> but as the cmake checks are already there present (find_path & find_library) I wouldn't remove them.

I can leave them, but that still would not solve the problem that we don't know the library version without pkgconfig.

fritsch commented:
I have nothing to say here. It's @wsnipex territory. If he says: good, it'll be good.

pkerling committed:
[cmake] Allow specifying version requirements for dependencies

pkerling committed:
[cmake][wayland] Deduplicate identical CMake file for BSD and Linux

pkerling committed:
[cmake] Reorganize Wayland detection

This now requires pkgconfig for correct operation. This should not be
a problem since wayland is only used on UNIX platforms anyway.

pkerling committed:
[cmake] Require pkgconfig for libxkbcommon

Since we will require a minimum libxkbcommon version, just finding the
headers is not enough and configuring will fail anyway if xkbcommon
was not found through pkgconfig.

pkerling committed:
[cmake] [wayland] Specify minimum version requirements for waylandpp and xkbcommon

waylandpp 0.1.5 for const-correct cursor functions
xkbcommon 0.4.1 for xkb_state_key_get_utf32

pkerling commented:
OK, sincerely hope this will be the last iteration. CMake version requirements implemented as discussed with @spiff and @wsnipex

pkerling commented:
jenkins build this please

pkerling committed:
Fix compiler warning

pkerling commented:
Jenkins is happy :-)

@wsnipex Can you review the new stuff, most importantly c26404b?
